neutral_generation/smart_convert.py

```diff
@@ -0,0 +1,284 @@
+import re
+import math
+import torch
+import spacy
+
+# SpaCy: lowercase is for dependency parser, uppercase is for part-of-speech tagger
+from spacy.symbols import nsubj, aux, conj, neg, VERB, AUX
+from pytorch_pretrained_bert import OpenAIGPTTokenizer, OpenAIGPTModel, OpenAIGPTLMHeadModel
+from constants import *
+
+nlp = spacy.load("en_core_web_sm")
+
+# Load pre-trained model (weights)
+model = OpenAIGPTLMHeadModel.from_pretrained('openai-gpt')
+model.eval()
+# Load pre-trained model tokenizer (vocabulary)
+tokenizer = OpenAIGPTTokenizer.from_pretrained('openai-gpt')
+
+
+def convert(s):
+    # input: SNAPE sentence (meaning 1 entity and 1 gender)
+    # output: sentence in gender-neutral form
+
+    # pluralize verbs: change all verbs in third-person singular to third-person plural
+    auxiliaries = identify_verbs_and_auxiliaries(s)
+    s = pluralize_verbs(s, auxiliaries)
+
+    # use regex to replace pronouns and gendered terms that have a clear mapping
+    SIMPLE_REPLACE = EASY_PRONOUNS
+    SIMPLE_REPLACE.update(GENDERED_TERMS)
+    for p, r in SIMPLE_REPLACE.items():
+        s = regex_token_replace(s, p, r)
+
+    # use a LM to break ties for pronouns when there is a one-to-many mapping
+    for p, choices in NON_FUNCTION_PRONOUNS.items():
+        s = smart_pronoun_replace(s, p, choices)
+
+    return s
+
+
+def regex_token_replace(s, old, new):
+    # replace occurrences of a token in a string with a new token
+    # TODO: can look into redoing replace_map with SpaCy (they save capitalization as well)
+    replace_map = [[old, new], [capitalize(old), capitalize(new)], [old.upper(), new.upper()]]
+
+    for j in range(len(replace_map)):
+        pattern = re.compile(r'\b{}\b'.format(replace_map[j][0]))
+        # TODO: add an exception to regex replacement when an apostrophe follows the pronoun
+        # e.g. He'shan is a location, but with this rule it would become They'shan
+        s = re.sub(pattern, replace_map[j][1], s)
+
+    return s
+
+
+def capitalize(word):
```

Comment:
I think Python has a built-in capitalize function that you can use.
See: https://www.geeksforgeeks.org/string-capitalize-python/

Reply:
Good to know. I'll remove this function.